cmd/gen-go-sample/inittree.go

```diff
@@ -185,71 +185,57 @@ func (t *initTree) index(k string) *initTree {
 }
 
 // Just do simple structs for now.
-// TODO(pongad): allow map and array index.
-//
-// spec = path [ '=' value ] .
-func (t *initTree) parseInit(txt string, info pbinfo.Info) error {
+// TODO(hzyi): allow map index.
+func (t *initTree) parseInit(path string, value string, info pbinfo.Info) error {
 	// The first ident is treated specially to be a member of the root.
 	// Since we know the root is a struct, a dot is the only legal token anyway,
 	// so just insert a dot here so we don't need to treat the first token specially.
-	sc, report := initScanner("." + txt)
+	sc, report := initScanner("." + path)
 
-	t, r, err := t.parsePathRest(sc, info)
+	t, _, err := t.parsePathRest(sc, info)
 	if err != nil {
 		return report(err)
 	}
-	if r != '=' {
-		return report(errors.E(nil, "expected '=', got %q", r))
+	if lv := t.leafVal; lv != "" {
+		return report(errors.E(nil, "value already set to %q", lv))
 	}
 
-	// TODO(pongad): handle resource names
-	switch r := sc.Scan(); r {
-	case scanner.Int, scanner.Float, scanner.String, scanner.Ident:
-		if lv := t.leafVal; lv != "" {
-			return report(errors.E(nil, "value already set to %q", lv))
+	if _, ok := t.typ.desc.(*descriptor.DescriptorProto); ok {
+		if value != "{}" {
+			return report(errors.E(nil, "invalid value for message: expecting `{}`, found %q", value))
 		}
-
-		tok := sc.TokenText()
-
-		if enum, ok := t.typ.desc.(*descriptor.EnumDescriptorProto); ok {
-			valid := false
-			for _, enumVal := range enum.Value {
-				if tok == enumVal.GetName() {
-					valid = true
-					break
-				}
-			}
-			if !valid {
-				return report(errors.E(nil, "invalid value for type %q: %q", enum.GetName(), tok))
-			}
-			t.typ.valFmt = enumFmt(info, enum)
-		} else {
-			pType := t.typ.prim
-			validPrim := validPrims[pType]
-			if validPrim == nil {
-				return report(errors.E(nil, "not a primitive type? %q", pType))
-			}
-			if !validPrim(tok) {
-				return report(errors.E(nil, "invalid value for type %q: %q", pType, tok))
-			}
-			if t.typ.prim == descriptor.FieldDescriptorProto_TYPE_BYTES {
-				t.typ.valFmt = bytesFmt()
+		return nil
+	} else if enum, ok := t.typ.desc.(*descriptor.EnumDescriptorProto); ok {
+		valid := false
+		for _, enumVal := range enum.Value {
+			if value == enumVal.GetName() {
+				valid = true
+				break
 			}
 		}
-		t.leafVal = tok
-
-	case '{':
-		if r := sc.Scan(); r != '}' {
-			return report(errors.E(nil, "bad format: expected '}', found %q", r))
+		if !valid {
+			return report(errors.E(nil, "invalid value for type %q: %q", enum.GetName(), value))
+		}
+		t.typ.valFmt = enumFmt(info, enum)
+	} else {
+		pType := t.typ.prim
+		validPrim := validPrims[pType]
+		if validPrim == nil {
+			return report(errors.E(nil, "not a primitive type? %q", pType))
+		}
+		if !validPrim(value) {
+			return report(errors.E(nil, "invalid value for type %q: %q", pType, value))
+		}
+		if t.typ.prim == descriptor.FieldDescriptorProto_TYPE_BYTES {
```

Comment:
`if pType ==`, we put `t.typ.prim` in a local variable already

Reply:
Done